grpc/internal/gateway/gateway.go

```diff
@@ -0,0 +1,110 @@
+package gateway
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/keratin/authn-server/app"
+	authnpb "github.com/keratin/authn-server/grpc"
+	"github.com/keratin/authn-server/server/sessions"
+	"google.golang.org/grpc/metadata"
+)
+
+type GatewayResponseMiddleware func(ctx context.Context, response http.ResponseWriter, m proto.Message) error
+
+func JSONMarshaler() runtime.Marshaler {
+	return new(customJSONMarshaler)
+}
+
+// StatusCodeMutator changes the HTTP Response status code to the desired mapping. The default mapping
+// by gRPC-Gateway doesn't have some of the desired responses (e.g. 201), which is why this function is needed.
+func StatusCodeMutator(ctx context.Context, response http.ResponseWriter, m proto.Message) error {
+	switch m.(type) {
+	case *authnpb.SignupResponseEnvelope, *authnpb.LoginResponseEnvelope, *authnpb.RefreshSessionResponseEnvelope, *authnpb.SubmitPasswordlessLoginResponseEnvelope, *authnpb.ChangePasswordResponseEnvelope:
+		response.WriteHeader(http.StatusCreated)
+	}
+	return nil
+}
+
+// CookieSetter extracts the session cookie from metadata and assigns it to a cookie. If the session
+// value is an empty string, then the cookie is marked to be removed.
+func CookieSetter(cfg *app.Config) GatewayResponseMiddleware {
+	return func(ctx context.Context, response http.ResponseWriter, m proto.Message) error {
+		switch m.(type) {
+		case *authnpb.LogoutResponse, *authnpb.SignupResponseEnvelope, *authnpb.LoginResponseEnvelope:
+			md, ok := runtime.ServerMetadataFromContext(ctx)
+			if !ok {
+				return fmt.Errorf("Failed to extract ServerMetadata from context")
+			}
+			ss := md.HeaderMD.Get(cfg.SessionCookieName)
+			if len(ss) != 1 {
+				return fmt.Errorf("Received more than a single session value")
+			}
+			sessions.Set(cfg, response, ss[0])
+		}
+		return nil
+	}
+}
+
+// CookieAnnotator reads the cookie from the *http.Request and sends it to the gRPC service as gRPC metadata
+func CookieAnnotator(app *app.App) func(ctx context.Context, req *http.Request) metadata.MD {
+	return func(ctx context.Context, req *http.Request) metadata.MD {
+		cookie, err := req.Cookie(app.Config.SessionCookieName)
+		if err != nil {
+			app.Reporter.ReportRequestError(err, req)
+			return nil
+		}
+		return metadata.MD{
+			app.Config.SessionCookieName: []string{cookie.Value},
+		}
+	}
+}
+
+// FormWrapper takes form values from application/x-www-form-urlencoded and converts it to JSON
+// Workaround adapted from: https://github.com/grpc-ecosystem/grpc-gateway/issues/7#issuecomment-358569373
+func FormWrapper(mux http.Handler) http.Handler {
```

Comment:
As I'm writing more tests, I found that this is imperfect at converting `application/x-www-form-urlencoded` to JSON. Due to `url.Values` storing slices of strings, then the value will always be string, and it'll fail if we're expecting anything other than string. I'll be brainstorming a better convertor.

There's [this](https://github.com/grpc-ecosystem/grpc-gateway/pull/548), but it's not merged, and needs to be evaluated. Basically, we need an adapter that takes in form values, and converts the value from `string` to the appropriate type.

Reply:
Ahh, right. The REST app normally manages type conversion in the handlers as part of the HTTP decoding process. Maybe we can encode simplifying assumptions to make it work for our limited needs? We mostly need to identify integers and booleans.

* integers: any param named `id`
* booleans: only bool currently is the `locked` param for account imports